lib/utils/log/slog_handler.go

```diff
@@ -0,0 +1,386 @@
+		case trace.Component:
+			const padding = trace.DefaultComponentPadding
+			component = fmt.Sprintf("[%v]", a.Value.String())
+			component = strings.ToUpper(padMax(component, padding))
+			if component[len(component)-1] != ' ' {
+				component = component[:len(component)-1] + "]"
+			}
+		case trace.ComponentFields:
+			switch fields := a.Value.Any().(type) {
+			case map[string]any:
+				for k, v := range fields {
+					s2.appendAttr(s2.preformatted, slog.Any(k, v))
+				}
+			case logrus.Fields:
+				for k, v := range fields {
+					s2.preformatted = s2.appendAttr(s2.preformatted, slog.Any(k, v))
+				}
+			}
+		default:
+			s2.preformatted = s2.appendAttr(s2.preformatted, a)
+		}
+	}
+	s2.component = component
+	return &s2
+}
+
+func (s *SlogTextHandler) appendUnopenedGroups(buf []byte) []byte {
+	for _, g := range s.unopenedGroups {
+		buf = fmt.Appendf(buf, "%s:", g)
+	}
+	return buf
+}
+
+// WithGroup opens a new group.
+func (s *SlogTextHandler) WithGroup(name string) slog.Handler {
+	if name == "" {
+		return s
+	}
+
+	s2 := *s
+	// Add an unopened group to h2 without modifying h.
+	s2.unopenedGroups = make([]string, len(s.unopenedGroups)+1)
+	copy(s2.unopenedGroups, s.unopenedGroups)
+	s2.unopenedGroups[len(s2.unopenedGroups)-1] = name
+	return &s2
+}
+
+// SlogJSONHandler is a [slog.Handler] that outputs messages in a json
+// format per the config file.
+type SlogJSONHandler struct {
+	handler *slog.JSONHandler
+}
+
+// NewSlogJSONHandler creates a SlogJSONHandler that outputs to w.
+func NewSlogJSONHandler(w io.Writer, level slog.Leveler) *SlogJSONHandler {
+	return &SlogJSONHandler{
+		handler: slog.NewJSONHandler(w, &slog.HandlerOptions{
+			AddSource: true,
+			Level:     level,
+			ReplaceAttr: func(groups []string, a slog.Attr) slog.Attr {
+				switch a.Key {
+				case trace.Component:
+					a.Key = "component"
+				case slog.LevelKey:
+					var level string
+					switch lvl := a.Value.Any().(slog.Level); lvl {
+					case TraceLevel:
+						level = "trace"
+					case slog.LevelDebug:
+						level = "debug"
+					case slog.LevelInfo:
+						level = "info"
+					case slog.LevelWarn:
+						level = "warning"
+					case slog.LevelError:
+						level = "error"
+					case slog.LevelError + 1:
+						level = "fatal"
+					default:
+						level = strings.ToLower(lvl.String())
+					}
+
+					a.Value = slog.StringValue(level)
+				case slog.TimeKey:
+					a.Key = "timestamp"
+					a.Value = slog.StringValue(a.Value.Time().Format(time.RFC3339))
+				case slog.MessageKey:
+					a.Key = "message"
+				case slog.SourceKey:
+					file, line := getCaller(a)
+					a = slog.String("caller", fmt.Sprintf("%s:%d", file, line))
+				}
+
+				return a
+			},
+		}),
+	}
+}
+
+func getCaller(a slog.Attr) (file string, line int) {
+	s := a.Value.Any().(*slog.Source)
+	count := 0
+	idx := strings.LastIndexFunc(s.File, func(r rune) bool {
+		if r == '/' {
+			count++
+		}
+
+		return count == 2
+	})
+	file = s.File[idx+1:]
+	line = s.Line
+
+	return
```

Comment:
Nit: It doesn't really matter but I really dislike naked returns - I find em super hard to read and they're fairly rare across our codebase.

```suggestion
	return file, line
```